src/main/java/com/google/api/gax/grpc/AbstractPage.java

```diff
@@ -0,0 +1,171 @@
+import com.google.api.gax.core.ApiFuture;
+import com.google.api.gax.core.ApiFutures;
+import com.google.api.gax.core.AsyncPage;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.Iterables;
+import java.util.Iterator;
+
+public abstract class AbstractPage<
+        RequestT,
+        ResponseT,
+        ResourceT,
+        PageT extends AbstractPage<RequestT, ResponseT, ResourceT, PageT>>
+    implements AsyncPage<ResourceT> {
+
+  private final PageContext<RequestT, ResponseT, ResourceT> context;
+  private final ResponseT response;
+
+  protected AbstractPage(PageContext<RequestT, ResponseT, ResourceT> context, ResponseT response) {
+    this.context = Preconditions.checkNotNull(context);
+    this.response = Preconditions.checkNotNull(response);
+  }
+
+  protected abstract PageT createPage(
+      PageContext<RequestT, ResponseT, ResourceT> context, ResponseT response);
+
+  @Override
+  public boolean hasNextPage() {
+    return !getNextPageToken().equals(context.pageDescriptor().emptyToken());
+  }
+
+  @Override
+  public String getNextPageToken() {
+    return context.pageDescriptor().extractNextToken(response);
+  }
+
+  @Override
+  public PageT getNextPage() {
+    if (hasNextPage()) {
+      RequestT nextRequest =
+          context.pageDescriptor().injectToken(context.request(), getNextPageToken());
+      ResponseT response =
+          ApiExceptions.callAndTranslateApiException(
+              context.callable().futureCall(context.request(), context.callContext()));
+      return createPage(context.withRequest(nextRequest), response);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public ApiFuture<PageT> getNextPageAsync() {
+    if (hasNextPage()) {
+      RequestT nextRequest =
+          context.pageDescriptor().injectToken(context.request(), getNextPageToken());
+      final PageContext<RequestT, ResponseT, ResourceT> newContext =
+          context.withRequest(nextRequest);
+      return ApiFutures.transform(
+          context.callable().futureCall(context.request(), context.callContext()),
+          new ApiFunction<ResponseT, PageT>() {
+            @Override
+            public PageT apply(ResponseT input) {
+              return createPage(newContext, input);
+            }
+          });
+    } else {
+      return ApiFutures.immediateFuture(null);
+    }
+  }
+
+  public PageT getNextPage(int pageSize) {
+    if (hasNextPage()) {
+      RequestT nextRequest =
+          context.pageDescriptor().injectToken(context.request(), getNextPageToken());
+      nextRequest = context.pageDescriptor().injectPageSize(nextRequest, pageSize);
+      ResponseT response =
+          ApiExceptions.callAndTranslateApiException(
+              context.callable().futureCall(nextRequest, context.callContext()));
+      return createPage(context.withRequest(nextRequest), response);
+    } else {
+      return null;
+    }
+  }
```

Comment:
Maybe have both `getNextPage()` and `getNextPage(int)` call a single private implementation to reduce duplication (`getNextPage(Integer)`)